File: java/Dijkstra.java
import java.util.*;

/**
 * Implements Dijkstra's shortest path algorithm.
 * Finds the shortest path between two nodes in a weighted graph
 * by always visiting the unvisited node with the smallest distance first.
 * Uses a PriorityQueue (min-heap) for efficient minimum extraction.
 */
public class Dijkstra {

    /**
     * Finds the shortest path from start to end using Dijkstra's algorithm.
     *
     * Steps:
     * 1. initialize all distances to infinity, except start node initialize it to 0
     * 2. Add start to priority queue sorted by distance (not coordinates)
     * 3. Poll the node with the smallest distance
     * 4. Skip if already visited, else mark the node as visited
     * 5. Stop if we reach to the end node
     * 6. Relax all unvisited neighbors (update if shorter path found)
     * 7. Recreating path using prev map from end to start
     *
     * @param graph the graph representing Kuwait's road network
     * @param start the starting area name
     * @param end   the destination area name
     * @return PathResult containing the shortest path, distance, and nodes visited
     */
    public static PathResult findPath(Graph graph, String start, String end) {
        PathResult result = new PathResult("Dijkstra");

        // dist: stores shortest known distance from start to each node
        Map<String, Double> dist = new HashMap<>();
        // prev: stores which node we came from (for path reconstruction)
        Map<String, String> prev = new HashMap<>();
        // visited: tracks nodes already processed to avoid revisiting
        Set<String> visited = new HashSet<>();

        // Initialize all distances to infinity and prev to null
        for (String n : graph.nodes()) {
            dist.put(n, Double.MAX_VALUE);
            prev.put(n, null);
        }
        // Start node has distance 0
        dist.put(start, 0.0);

        // Priority queue sorted by distance - always polls the closest node
        PriorityQueue<String> pq = new PriorityQueue<>(
                Comparator.comparingDouble(n -> dist.getOrDefault(n, Double.MAX_VALUE)));
        pq.add(start);

        // Main loop: process nodes until queue is empty or destination reached
        while (!pq.isEmpty()) {
            // Poll the node with the smallest distance
            String u = pq.poll();

            // Skip if already visited (handles duplicate entries in PQ)
            if (visited.contains(u)) {
                continue;
            }
            // Mark as visited so we don't process it again
            visited.add(u);

            // Early termination: destination reached
            if (u.equals(end)) {
                break;
            }

            // Relaxation: check all neighbors of current node
            for (Edge e : graph.neighbors(u)) {
                // Skip already visited neighbors
                if (visited.contains(e.dest)) {
                    continue;
                }
                // Calculate new distance through current node
                double alt = dist.get(u) + e.weight;
                // If shorter path found, update distance and predecessor
                if (alt < dist.get(e.dest)) {
                    dist.put(e.dest, alt);
                    prev.put(e.dest, u);
                    pq.add(e.dest);
                }
            }
        }

        // Store number of nodes visited
        result.nodesVisited = visited.size();

        // Path reconstruction: backtrack from end to start using prev map
        if (dist.get(end) < Double.MAX_VALUE) {
            result.totalDist = dist.get(end);
            String n = end;
            // Add each node to the front of the list so path reads start -> end
            while (n != null) {
                result.path.add(0, n);
                n = prev.get(n);
            }
        }
        return result;
    }
}

File: lib/dijkstra.ts
import { AlgoStep, GraphData, PQEntry } from "./types";
import { getNeighbors } from "./graphData";

function reconstructPath(prev: Record<string, string | null>, end: string): string[] {
  const path: string[] = [];
  let n: string | null = end;
  while (n !== null && n !== undefined) {
    path.unshift(n);
    n = prev[n] ?? null;
  }
  return path;
}

export function runDijkstra(graph: GraphData, start: string, end: string): AlgoStep[] {
  const steps: AlgoStep[] = [];
  const INF = Infinity;

  const dist: Record<string, number> = {};
  const prev: Record<string, string | null> = {};
  const visited: Set<string> = new Set();

  for (const node of graph.nodes) {
    dist[node.id] = INF;
    prev[node.id] = null;
  }
  dist[start] = 0;

  // Min-heap simulation using a sorted array (for clarity in visualization)
  let pq: PQEntry[] = [{ node: start, priority: 0 }];

  let stepNum = 0;

  steps.push({
    stepNumber: stepNum++,
    currentNode: start,
    visited: [],
    inQueue: [start],
    pq: [{ node: start, priority: 0 }],
    scores: { ...dist },
    prev: { ...prev },
    updatedNeighbors: [],
    explanation: `Initialize: Set distance to "${start}" = 0. All other nodes = ∞. Add "${start}" to the Priority Queue.`,
    concept: "initialization",
  });

  while (pq.length > 0) {
    pq.sort((a, b) => a.priority - b.priority);
    const { node: u } = pq.shift()!;

    if (visited.has(u)) continue;
    visited.add(u);

    if (u === end) {
      const path = reconstructPath(prev, end);
      steps.push({
        stepNumber: stepNum++,
        currentNode: u,
        visited: [...visited],
        inQueue: pq.map((e) => e.node),
        pq: [...pq],
        scores: { ...dist },
        prev: { ...prev },
        updatedNeighbors: [],
        finalPath: path,
        explanation: `Reached destination "${end}"! Distance = ${dist[end]} km. Path: ${path.join(" → ")}`,
        concept: "done",
      });
      break;
    }

    const neighbors = getNeighbors(u, graph.edges);
    const updatedNeighbors = [];

    for (const { dest, weight } of neighbors) {
      if (visited.has(dest)) continue;
      const alt = dist[u] + weight;
      if (alt < dist[dest]) {
        const oldDist = dist[dest];
        dist[dest] = alt;
        prev[dest] = u;
        pq.push({ node: dest, priority: alt });
        updatedNeighbors.push({ node: dest, oldScore: oldDist, newScore: alt });
      }
    }

    const sortedPQ = [...pq].sort((a, b) => a.priority - b.priority);

    steps.push({
      stepNumber: stepNum++,
      currentNode: u,
      visited: [...visited],
      inQueue: sortedPQ.map((e) => e.node),
      pq: sortedPQ,
      scores: { ...dist },
      prev: { ...prev },
      updatedNeighbors,
      explanation: buildDijkstraExplanation(u, dist[u], updatedNeighbors),
      concept: "relax",
    });
  }

  return steps;
}

function buildDijkstraExplanation(
  node: string,
  dist: number,
  updates: Array<{ node: string; oldScore: number; newScore: number }>
): string {
  let msg = `Dequeue "${node}" (dist=${dist} km) — mark as visited. `;
  if (updates.length === 0) {
    msg += "No neighbors were improved.";
  } else {
    const parts = updates.map((u) => {
      const old = u.oldScore === Infinity ? "∞" : `${u.oldScore}`;
      return `"${u.node}": ${old} → ${u.newScore}`;
    });
    msg += `Relax edges: ${parts.join(", ")}.`;
  }
  return msg;
}
